pynumaflow/function/server.py

```diff
@@ -11,55 +12,90 @@
 from pynumaflow._constants import (
     FUNCTION_SOCK_PATH,
     DATUM_KEY,
+    WIN_START_TIME,
+    WIN_END_TIME,
     MAX_MESSAGE_SIZE,
 )
-from pynumaflow.function import Messages, Datum
+from pynumaflow.function import Messages, Datum, IntervalWindow, Metadata
 from pynumaflow.function.generated import udfunction_pb2
 from pynumaflow.function.generated import udfunction_pb2_grpc
 from pynumaflow.types import NumaflowServicerContext
 
+
 if os.getenv("PYTHONDEBUG"):
     logging.basicConfig(level=logging.DEBUG)
 
 _LOGGER = logging.getLogger(__name__)
 
 UDFMapCallable = Callable[[str, Datum], Messages]
+UDFReduceCallable = Callable[[str, Iterator[Datum], Metadata], Messages]
 _PROCESS_COUNT = multiprocessing.cpu_count()
 MAX_THREADS = int(os.getenv("MAX_THREADS", 0)) or (_PROCESS_COUNT * 4)
 
 
+def default_map_handler(_: str, __: Datum) -> Messages:
+    return
```

Comment:
I want to make input arguments map_handler and reduce_handler to be optional.
That's why I added these two `default_handler` functions.

Need suggestions.. whether there's a more standard way to solve this?

Reply:
I don't think you can do that at compile time. Python's type checking system is not that mature. @ab93 ?

Reply:
> I don't think you can do that at compile time. Python's type checking system is not that mature. @ab93 ?

Tested and passed. Does it mean compile succeeded? https://github.com/numaproj/numaflow-python/pull/40#issue-1500993188

Reply:
if I call both reduce function and map function in the above test
```go
response, err := c.ReduceFn(ctx, reduceCh)
fmt.Println(response, err)
response, err = c.MapFn(ctx, testDatum)
fmt.Println(response, err)
```
got
```sh
[key:"client" value:"counter:10 interval_window_start:2022-09-12 16:00:00+00:00 interval_window_end:2022-09-12 16:01:00+00:00"] <nil>
[] failed to execute c.grpcClt.MapFn(): rpc error: code = Unknown desc = Exception calling application: 'NoneType' object has no attribute 'items'
```

Reply:
so, if the user passes the wrong handler, we can catch that only a runtime.

Reply:
Wrong handler?
You mean user create a reduce udf but use it as a map udf? Then yes the error can only be triggered when the apply udf is called

Reply:
yes